eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs

```diff
@@ -271,65 +257,102 @@ actionWellFormed ga = failureUnless isWellFormed $ MalformedProposal ga
+  GovActionState
+    { gasId = actionId
+    , gasCommitteeVotes = mempty
+    , gasDRepVotes = mempty
+    , gasStakePoolVotes = mempty
+    , gasDeposit = deposit
+    , gasReturnAddr = returnAddress
+    , gasAction = action
+    , gasProposedIn = curEpoch
+    , gasExpiresAfter = curEpoch + expiry
+    , gasChildren = mempty
+    }
+
+-- | Add a child to the already enacted actions. Returns a `Just
+-- (PrevGovActionIdsChildren era)` with the child added, or `Nothing` if
+-- the `PrevGovActionId` is not found.
+--
+-- NOTE:
+-- If both the already enacted action as well as the previous Id
+-- referenced by the action being addecd are `SNothing` the checks pass
+addToPrevGovActionIdsChildren ::
+  GovActionId (EraCrypto era) ->
+  GovAction era ->
   PrevGovActionIds era ->
-  ProposalsSnapshot era ->
-  GovProcedures era ->
-  Test (ConwayGovPredFailure era)
-checkProposalsHaveAValidPrevious prevGovActionIds snapshots procedures =
-  let isValidPrevGovActionId ::
-        StrictMaybe (PrevGovActionId p (EraCrypto era)) ->
-        (GovAction era -> Bool) ->
-        Bool
-      isValidPrevGovActionId prevGovActionId snapshotCond =
-        case prevGovActionId of
-          -- The case of having an SNothing as valid, for the very first proposal ever, is handled in `prevActionAsExpected`
-          SNothing -> False
-          SJust (PrevGovActionId govActionId) ->
-            case snapshotLookupId govActionId $ snapshots of
-              Nothing -> False
-              -- lookup has to succeed _and_ purpose of looked-up action has to match condition
-              Just found -> snapshotCond $ gasAction found
-      isValid proposal =
-        prevActionAsExpected (proposal ^. pProcGovActionL) prevGovActionIds
-          || case proposal ^. pProcGovActionL of
-            ParameterChange prev _ ->
-              isValidPrevGovActionId prev $ \case ParameterChange {} -> True; _ -> False
-            HardForkInitiation prev _ ->
-              isValidPrevGovActionId prev $ \case HardForkInitiation {} -> True; _ -> False
-            TreasuryWithdrawals _ -> True
-            NoConfidence prev ->
-              isValidPrevGovActionId prev $ \case NoConfidence {} -> True; UpdateCommittee {} -> True; _ -> False
-            UpdateCommittee prev _ _ _ ->
-              isValidPrevGovActionId prev $ \case NoConfidence {} -> True; UpdateCommittee {} -> True; _ -> False
-            NewConstitution prev _ ->
-              isValidPrevGovActionId prev $ \case NewConstitution {} -> True; _ -> False
-            InfoAction -> True
-      invalidProposals = Seq.filter (not . isValid) $ SSeq.fromStrict $ OSet.toStrictSeq $ procedures ^. govProceduresProposalsL
-   in failureUnless (Seq.null invalidProposals) $ InvalidPrevGovActionIdsInProposals invalidProposals
+  PrevGovActionIdsChildren era ->
+  Maybe (PrevGovActionIdsChildren era)
+addToPrevGovActionIdsChildren actionId action PrevGovActionIds {..} children =
```

Comment:
This function resembles this one: https://github.com/input-output-hk/cardano-ledger/blob/2f29af846d743d1c79e77ab52c2142ec1a127063/eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs#L343

I wonder if we can reduce duplication and extract a common part between the two and reuse it somehow in both places: here and in RATIFY rule

Reply:
May I tackle this in a follow up PR with the test coverage?